Feature: async mint/redeem

JuampiRombola opened the pull request.

JuampiRombola committed:
chore: wip async mint

JuampiRombola committed:
chore: integrate sign, submit and get status of async mint

JuampiRombola committed:
chore: wip redirect to manual mint when collaterals are acquired

JuampiRombola committed:
feat: add automated sub navigation menu

JuampiRombola committed:
chore: restore zap-mint/index

JuampiRombola committed:
chore: restore zap-mint/atom.ts

JuampiRombola committed:
chore: split automated swaps

JuampiRombola committed:
feat: add gnosis safe required modal

JuampiRombola committed:
chore: change navigation side panel

JuampiRombola committed:
chore: change gnosis safe required component

JuampiRombola committed:
chore: change async mint styles

JuampiRombola committed:
chore: add approve & mint button

JuampiRombola committed:
chore: add success msg to async mint

JuampiRombola committed:
fix: collaterals height

JuampiRombola committed:
chore: wip async redeem

JuampiRombola committed:
feat: async redeem

JuampiRombola committed:
chore: restore constant

lcamargof commented on src/hooks/useAsyncSwap.ts:
```diff
@@ -0,0 +1,92 @@
+  )
+
+  useEffect(() => {
+    setAsyncSwapEndpoint(endpoint ?? '')
```
why? is only used here, and if its going to be used on other module, then coupling it in this hook would be weird as this hooks is serving an specific purpose

JuampiRombola replied:
I'm going to remove it. I was actually leaving it because in Zaps we use it to send to Mixpanel as the event ID (and to extract all the sent parameters from there). I figured we'd need it here too, but for now I'm removing it.

lcamargof commented on src/hooks/useAsyncSwap.ts:
```diff
@@ -0,0 +1,92 @@
+      const response = await fetch(endpoint)
+      if (!response.ok) {
+        const error = response.status
+        mixpanel.track('async-zap', {
```
move that mixpanel body to a common const "analyticsMeta" to reuse it for success/error

lcamargof commented on src/state/chain/updaters/AtomUpdater.tsx:
```diff
@@ -57,9 +60,30 @@ const AtomUpdater = () => {
         wa: account.address,
       })
+      // Check if the wallet is a Safe Multisig
+      const checkIfSafe = async () => {
```
move this outside of the useEffect like the timestamp fn, and only run it on a different useEffect that only checks for account?.address change, currently this re-run even if they change chains

lcamargof commented on src/utils/constants.ts:
```diff
@@ -22,7 +22,7 @@ export const RESERVE_FORUM = 'https://forum.reserve.org/'
   'https://reserve.canny.io/defi-surfaces-bug-reporting'
 export const RESERVE_STORAGE = 'https://storage.reserve.org/'
-export const RESERVE_API = 'https://api.reserve.org/'
+export const RESERVE_API = 'https://api-staging.reserve.org/'
```
Add a TODO or add this as an environment variable, this can be merged into prod by mistake

lcamargof commented on src/views/index-dtf/issuance/async-swaps/async-mint/index.tsx:
```diff
@@ -0,0 +1,152 @@
+
+  return (
+    <div className="flex flex-col h-full">
+      <Swap
```
Replace the "Swap" component for a component tree on which each component talk with the specific state they care about, current component is too big

lcamargof commented on src/views/index-dtf/issuance/async-swaps/async-mint/index.tsx:
```diff
@@ -0,0 +1,152 @@
+        }}
+        loading={isLoading || loadingAfterRefetch}
+        disabled={orderSubmitted}
+        classNameInput={cn(
```
seems like we need an input variant rather than explicit class names on props

lcamargof commented on src/views/index-dtf/issuance/async-swaps/async-redeem/index.tsx:
```diff
@@ -0,0 +1,184 @@
+          onChange: setInputAmount,
+          onMax,
+        }}
+        customInput={
```
use component composition instead of prop

lcamargof commented on src/views/index-dtf/issuance/async-swaps/async-redeem/index.tsx:
```diff
@@ -0,0 +1,184 @@
+          onMax,
+        }}
+        customInput={
+          Object.keys(redeemAssets).length > 0 ? (
```
Create specific component

lcamargof commented on src/views/index-dtf/issuance/async-swaps/collaterals.tsx:
```diff
@@ -0,0 +1,135 @@
+      setTimeout(() => setIsVisible(true), 0)
+    } else {
+      setIsVisible(false)
+      setTimeout(() => setShouldRender(false), 300)
```
clear timeouts on unmount, they could set state on an unmounted component

lcamargof commented on src/views/index-dtf/issuance/async-swaps/collaterals.tsx:
```diff
@@ -0,0 +1,135 @@
+        isVisible ? 'w-[400px]' : 'w-0'
+      )}
+    >
+      {cowswapOrders.map(({ orderId, quote, status }) => (
```
Create a CowSwapOrder component

lcamargof commented on src/views/index-dtf/issuance/async-swaps/collaterals.tsx:
```diff
@@ -0,0 +1,135 @@
+              </div>
+            </div>
+          </div>
+          <div
```
OrderStatus component

lcamargof commented on src/views/index-dtf/issuance/async-swaps/index.tsx:
```diff
@@ -0,0 +1,183 @@
+
+  if (isSafeMultisig) {
+    return (
+      <div className="container flex flex-col items-center sm:justify-start md:justify-center gap-2 lg:border-2 lg:border-secondary lg:bg-secondary/30 lg:h-[calc(100vh-100px)] dark:bg-card rounded-4xl w-full">
```
the same container div is used 3 times, create a Container component with the variations instead

lcamargof commented on src/views/index-dtf/issuance/async-swaps/success.tsx:
```diff
@@ -0,0 +1,344 @@
+    return <Transactions />
+  }
+
+  return (
```
split into multiple components

lcamargof reviewed (commented):
Refactor big components into smaller chunks for better readability and maintenance.

JuampiRombola committed:
chore: refactor swap component

JuampiRombola committed:
chore: add variants to Switch

JuampiRombola commented on src/app-routes.tsx:
```diff
@@ -81,6 +82,7 @@ const AppRoutes = () => (
         path={`${ROUTES.ISSUANCE}/manual`}
         element={<IndexDTFManualIssuance />}
       />
+      <Route path={`${ROUTES.ISSUANCE}/automated`} element={<AsyncSwaps />} />
```
When reviewing the designs with Akshat, we agreed that "Automated" best reflects this type of operation. That's why the menu also says "Automated"

<img width="173" alt="image" src="https://github.com/user-attachments/assets/68ca869a-7bae-4fb7-bbf8-846aca267a20" />

JuampiRombola committed:
chore: refactor useAsyncSwap

JuampiRombola committed:
chore: remove asyncSwapEndpointAtom

JuampiRombola committed:
chore: add commonAnalytics

JuampiRombola committed:
chore: clear timeouts on unmount

JuampiRombola committed:
chore: add cowswap component

JuampiRombola committed:
chore: add OrderStatus component

JuampiRombola committed:
chore: refactor success component

JuampiRombola committed:
fix: collaterals height

JuampiRombola committed:
wip: integration

JuampiRombola committed:
chore: fix build

JuampiRombola committed:
chore: working multi signature

JuampiRombola committed:
chore: wip async mint integration

JuampiRombola committed:
fix: success component

JuampiRombola committed:
chore: working async mint w/ safe

JuampiRombola committed:
chore: async redeem

JuampiRombola committed:
chore: clean atoms

JuampiRombola committed:
feat: add insufficient balance

JuampiRombola committed:
fix: redeem success msg

JuampiRombola committed:
chore: reset atoms

JuampiRombola committed:
feat: add failed order retry

JuampiRombola committed:
chore: quote config update

JuampiRombola committed:
chore: cowswap link always visible

JuampiRombola committed:
fix: redeem output and cowswap component

JuampiRombola committed:
chore: handle USDC edge case

JuampiRombola committed:
feat: add async swap settings

JuampiRombola committed:
chore: add use wallet balance config

JuampiRombola committed:
chore: add async mint details

JuampiRombola committed:
chore: universal useQuote

JuampiRombola committed:
chore: universal useQuoteSignature

JuampiRombola committed:
chore: refactor

JuampiRombola committed:
chore: refactor

JuampiRombola committed:
chore: add rate limit and universal statuses

JuampiRombola committed:
fix: collaterals view

JuampiRombola committed:
chore: working universal mint

JuampiRombola committed:
chore: universal refresh

JuampiRombola committed:
fix: refresh

JuampiRombola committed:
chore: refactor async mint/redeem

JuampiRombola committed:
chore: failed universal orders

JuampiRombola committed:
chore: remove universal failed order component

JuampiRombola committed:
fix: styles

JuampiRombola committed:
chore: redeem

JuampiRombola committed:
fix: unintentional refersh

JuampiRombola committed:
fix: unintentional refersh & icons

JuampiRombola committed:
fix: flaky mint event listening

JuampiRombola committed:
chore: adjust styles

JuampiRombola committed:
chore: add retries

JuampiRombola committed:
chore: add info and warning msg

JuampiRombola committed:
Merge branch 'master' into feature/async-zap

JuampiRombola committed:
fix: add missing key on map

JuampiRombola committed:
chore: update package-lock.json

JuampiRombola committed:
chore: pick best quote between cowswap & universal

JuampiRombola committed:
chore: improve build

JuampiRombola committed:
chore: improve balance after and before

JuampiRombola committed:
chore: add appData and appDataHex to cowswap

JuampiRombola committed:
chore: move universal to reserver-api

JuampiRombola committed:
chore: change success view styles

JuampiRombola committed:
chore: increase buffer

JuampiRombola committed:
fix: rerenders

JuampiRombola committed:
chore: change gnosis safe required to atomic batch required

JuampiRombola committed:
chore: change universal api base url

JuampiRombola committed:
Merge branch 'master' into feature/async-zap

JuampiRombola committed:
chore: remove unused file

JuampiRombola committed:
chore: change build command

JuampiRombola committed:
chore: update package.json

JuampiRombola committed:
chore: add useAtomicBatch hook

JuampiRombola committed:
chore: update cowswap packages

JuampiRombola committed:
Merge branch 'master' into feature/async-zap

JuampiRombola committed:
fix: atomic support

JuampiRombola committed:
chore: improve wallet connection

lcamargof committed:
chore: merge master into feature/async-zap

Resolve conflicts: use master's dep versions, simple wallet chunk,
build scripts. Keep cowprotocol libs (bumped to latest), universal-sdk,
node-fetch polyfill. Drop theme-ui.

lcamargof committed:
feat: new design imp

lcamargof committed:
feat: happy path

lcamargof committed:
chore: disable top up if user doesnt hold balance

lcamargof committed:
chore: add cc context

akshatmittal committed:
Merge branch 'master' into feature/async-zap

# Conflicts:
#	package-lock.json
#	package.json
